src/relations/tls.py

```diff
@@ -0,0 +1,297 @@
+
+    @property
+    def certificate_authority(self) -> str:
+        """The TLS certificate authority"""
+        return self._secrets.get_value(relations.secrets.UNIT_SCOPE, _TLS_CA)
+
+    def save_certificate(self, event: tls_certificates.CertificateAvailableEvent) -> None:
+        """Save TLS certificate in peer relation unit databag."""
+        if (
+            event.certificate_signing_request.strip()
+            != self._secrets.get_value(relations.secrets.UNIT_SCOPE, _TLS_REQUESTED_CSR).strip()
+        ):
+            logger.warning("Unknown certificate received. Ignoring.")
+            return
+        if (
+            self.certificate_saved
+            and event.certificate_signing_request.strip()
+            == self._secrets.get_value(relations.secrets.UNIT_SCOPE, _TLS_ACTIVE_CSR)
+        ):
+            # Workaround for https://github.com/canonical/tls-certificates-operator/issues/34
+            logger.debug("TLS certificate already saved.")
+            return
+        logger.debug(f"Saving TLS certificate {event=}")
+        self._secrets.set_value(relations.secrets.UNIT_SCOPE, _TLS_CERTIFICATE, event.certificate)
+        self._secrets.set_value(relations.secrets.UNIT_SCOPE, _TLS_CA, event.ca)
+        self._secrets.set_value(relations.secrets.UNIT_SCOPE, _TLS_CHAIN, json.dumps(event.chain))
+        self._secrets.set_value(
+            relations.secrets.UNIT_SCOPE,
+            _TLS_ACTIVE_CSR,
+            self._secrets.get_value(relations.secrets.UNIT_SCOPE, _TLS_REQUESTED_CSR),
+        )
+        logger.debug(f"Saved TLS certificate {event=}")
+        self._charm.reconcile(event=None)
+
+    def _generate_csr(self, key: bytes) -> bytes:
+        """Generate certificate signing request (CSR)."""
+        unit_name = self._charm.unit.name.replace("/", "-")
+        extra_hosts, extra_ips = self._charm.get_all_k8s_node_hostnames_and_ips()
+        return tls_certificates.generate_csr(
+            private_key=key,
+            subject=socket.getfqdn(),
+            organization=self._charm.app.name,
+            sans_dns=[
+                unit_name,
+                f"{unit_name}.{self._charm.app.name}-endpoints",
+                f"{unit_name}.{self._charm.app.name}-endpoints.{self._charm.model_service_domain}",
+                f"{self._charm.app.name}-endpoints",
+                f"{self._charm.app.name}-endpoints.{self._charm.model_service_domain}",
+                f"{unit_name}.{self._charm.app.name}",
+                f"{unit_name}.{self._charm.app.name}.{self._charm.model_service_domain}",
+                self._charm.app.name,
+                f"{self._charm.app.name}.{self._charm.model_service_domain}",
+                *extra_hosts,
+            ],
+            sans_ip=[
+                str(self._charm.model.get_binding("juju-info").network.bind_address),
+                "127.0.0.1",
+                *extra_ips,
+            ],
+        )
```

Comment:
The content of the CSR seems quite large, particularly having the `127.0.0.1` and internal IP addresses in there. This will limit the `tls-providers` that can be used with this charm. In particular, this will prevent the use of all ACME providers.

You might want to make this configurable in some ways depending on the deployments scenarios you expect.